### ISSUE TITLE
Does zenpy support pagination

### ISSUE THREAD

**Issue by caduri** [facetoe/zenpy#158]
Hi @facetoe I'm trying to find a way to use pagination using zenpy
according to Zendesk API they support pagination using the `per_page` and `page` params

Im using this function `zenpy.views.tickets(view_id)`

Thanks in advance

**Comment by facetoe** [comment]
Hi @caduri, yes Zenpy supports pagination transparently via generators, eg:

```python
>>> type(zenpy_client.views.tickets(view_id))
<class 'zenpy.lib.generator.ViewResultGenerator'>
```
You just need to iterate over the returned generator until it is exhausted:

```python
for ticket in zenpy_client.views.tickets(view=view_id):
    print(ticket)
```

Hope this helps!

**Comment by caduri** [comment]
Thanks @facetoe for your reply.

The generator is very helpful when you need all the results, but what I'm asking is exactly the opposite, I need a way to get only 1 page and also to set the results per page

I think you support it in the search functionality

**Comment by facetoe** [comment]
Ahh I see what you mean. No this is not currently supported. I'm wondering if something like the following would suffice:

```python
max = 100
for i, ticket in enumerate(zenpy_client.views.tickets(view=view_id)):
    if i > max:
        break
    print(ticket)
```

This would allow you to only retrieve the first N objects. If this won't work for your use case could you tell me a little more about what you are trying to achieve?

**Comment by facetoe** [comment]
Yes that's true, not ideal. I can look at adding support for this, although it might be a bit tricky.

**Comment by caduri** [comment]
I think about adding kwargs as a parameter to views.tickets and forward it to BaseApi._get
and there add it to url. It will look like this

`
def _get(self, url, raw_response=False, **kwargs):
        if kwargs:
            params = '&'.join("%s=%r" % (key, val) for (key, val) in kwargs.iteritems())
            url = '?'.join([url, params])
            response = self._call_api(self.session.get, url, timeout=self.timeout)
        else:
            response = self._call_api(self.session.get, url, timeout=self.timeout)

        if raw_response:
            return response
        else:
            return self._process_response(response)
`

What do you think?

**Comment by caduri** [comment]
This will work in my use case

**Comment by facetoe** [comment]
I think if we add support for this it needs to work anywhere a generator is returned. Currently the `Endpoint` classes are responsible for building URLs, so that's one logical place this could be done. Another option could be to pass these parameters to the `requests` HTTP method as `params`. I'll need to stare at the code a bit and figure out a way to add this cleanly so it works everywhere you would expect it to work without causing any regressions.

Ideally it would be worked into the `Endpoint` classes somehow so we keep all that logic in the same place.

**Comment by caduri** [comment]
Yeah I figured this is what you'll say. The wrapper is too generic for me to do a quick PR

Let me know if I can help

**Comment by caduri** [comment]
Any update on the subject?

**Comment by facetoe** [comment]
Hi @caduri, no I haven't had a chance to look into this yet. I'm a bit busy at the moment so it might take a while until I get the chance unfortunately.

**Comment by facetoe** [comment]
Hi @caduri, 

I've added a prototype implementation of slice based pagination. The current implementation has some limitations:

* Does not support indexing at all
* Does not support setting `per_page`
* Does not support negative values
* Always pulls the first 100 objects
* Does not support multiple accesses

On the bright side, it has a few advantages:

* Supports simple use cases, like retrieving ticket 720 to 750  efficiently (two API calls, could be optimized to one)
* The implementation is isolated from the rest of the Zenpy code, which should make it easier for people to hack on without worrying about breaking things
* All of the above limitations could theoretically be overcome, making it really awesome and powerful

Example usage:
```python
ticket_generator = zenpy_client.tickets()
tickets = ticket_generator[720:750]
print(tickets)
```

You can try it out here - https://github.com/facetoe/zenpy/tree/support_pagination.

**Comment by facetoe** [comment]
Now supports passing `page_size` as the slice "step" parameter.

**Comment by facetoe** [comment]
Merged in b9b5479cad61e861bba6abd7e63700036591e08f
